riscv: Provide a vector only implementation of CHACHA20 cipher

cyyself opened the pull request:
Although we have a Zvkb version of Chacha20, the Zvkb from the RISC-V Vector Cryptography Bit-manipulation extension was ratified in late 2023 and does not come to the RVA23 Profile. Many CPUs in 2024 currently do not support Zvkb but may have Vector and Scalar Bit-manipulation, which are already in the RVA22 Profile. This commit provides a vector-only implementation that replaced the vror with vsll+vsrl+vor and can provide enough speed for Chacha20 for new CPUs this year.

##### Performance Evaluation

Platform: Canaan Kendryte K230
CPU: T-Head C908 @1.6GHz
Compiler: gcc version 13.2.1 20230801 (GCC)

```console
[cyy@archlinux ssl]$ neofetch
                   -`                    cyy@archlinux
                  .o+`                   -------------
                 `ooo/                   OS: Arch Linux riscv64
                `+oooo:                  Host: Canaan Kendryte K230
               `+oooooo:                 Kernel: 6.8.0+
               -+oooooo+:                Uptime: 4 days, 17 hours, 27 mins
             `/:-:++oooo+:               Packages: 156 (pacman)
            `/++++/+++++++:              Shell: bash 5.2.26
           `/++++++++++++++:             Terminal: /dev/pts/0
          `/+++ooooooooooooo/`           CPU: (1)
         ./ooosssso++osssssso+`          Memory: 72MiB / 477MiB
        .oossssso-````/ossssss+`
       -osssssso.      :ssssssso.
      :osssssss/        osssso+++.
     /ossssssss/        +ssssooo/-
   `/ossssso+/:-        -:/+osssso+-
  `+sso+:-`                 `.-/+oso:
 `++:.                           `-/+/
 .`                                 `/

[cyy@archlinux ssl]$ cat /proc/cpuinfo
processor	: 0
hart		: 0
isa		: rv64imafdcv_zicbom_zicntr_zicsr_zifencei_zihpm_zba_zbb_zbc_zbs_svpbmt
mmu		: sv39
mvendorid	: 0x5b7
marchid		: 0x8000000009140d00
mimpid		: 0x50000
hart isa	: rv64imafdcv_zicbom_zicntr_zicsr_zifencei_zihpm_zba_zbb_zbc_zbs_svpbmt

[cyy@archlinux ssl]$ unset OPENSSL_riscvcap
[cyy@archlinux ssl]$ ./bin/openssl speed -evp chacha20
Doing ChaCha20 ops for 3s on 16 size blocks: 9190817 ChaCha20 ops in 2.90s
Doing ChaCha20 ops for 3s on 64 size blocks: 3541082 ChaCha20 ops in 2.98s
Doing ChaCha20 ops for 3s on 256 size blocks: 998612 ChaCha20 ops in 2.98s
Doing ChaCha20 ops for 3s on 1024 size blocks: 258575 ChaCha20 ops in 2.99s
Doing ChaCha20 ops for 3s on 8192 size blocks: 32655 ChaCha20 ops in 2.98s
Doing ChaCha20 ops for 3s on 16384 size blocks: 16335 ChaCha20 ops in 2.99s
version: 3.4.0-dev
built on: Fri Apr 19 06:03:02 2024 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
CPUINFO: N/A
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
ChaCha20         50707.96k    76050.08k    85786.80k    88555.45k    89768.38k    89509.24k
[cyy@archlinux ssl]$ export OPENSSL_riscvcap=rv64gc_v_zba_zbb
[cyy@archlinux ssl]$ ./bin/openssl speed -evp chacha20
Doing ChaCha20 ops for 3s on 16 size blocks: 9400208 ChaCha20 ops in 2.97s
Doing ChaCha20 ops for 3s on 64 size blocks: 3561831 ChaCha20 ops in 2.99s
Doing ChaCha20 ops for 3s on 256 size blocks: 1417570 ChaCha20 ops in 2.98s
Doing ChaCha20 ops for 3s on 1024 size blocks: 370833 ChaCha20 ops in 2.98s
Doing ChaCha20 ops for 3s on 8192 size blocks: 57722 ChaCha20 ops in 2.99s
Doing ChaCha20 ops for 3s on 16384 size blocks: 28833 ChaCha20 ops in 2.98s
version: 3.4.0-dev
built on: Fri Apr 19 06:03:02 2024 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -Wa,--noexecstack -Wall -O3 -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
CPUINFO: N/A
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
ChaCha20         50640.85k    76239.86k   121777.83k   127427.18k   158146.70k   158523.45k
[cyy@archlinux ssl]$ unset OPENSSL_riscvcap
[cyy@archlinux ssl]$ ./bin/openssl speed -evp chacha20
Doing ChaCha20 ops for 3s on 16 size blocks: 10710666 ChaCha20 ops in 2.99s
Doing ChaCha20 ops for 3s on 64 size blocks: 4656347 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 256 size blocks: 1390592 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 1024 size blocks: 366078 ChaCha20 ops in 3.01s
Doing ChaCha20 ops for 3s on 8192 size blocks: 46493 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 16384 size blocks: 23223 ChaCha20 ops in 3.00s
version: 3.4.0-dev
built on: Fri Apr 19 06:13:13 2024 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -Wa,--noexecstack -Wall -O3 -march=rv64gcv_zba_zbb -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
CPUINFO: N/A
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
ChaCha20         57314.60k    99335.40k   118663.85k   124539.49k   126956.89k   126828.54k
[cyy@archlinux ssl]$ export OPENSSL_riscvcap=rv64gc_v_zba_zbb
[cyy@archlinux ssl]$ ./bin/openssl speed -evp chacha20
Doing ChaCha20 ops for 3s on 16 size blocks: 11064374 ChaCha20 ops in 2.96s
Doing ChaCha20 ops for 3s on 64 size blocks: 4723040 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 256 size blocks: 1426926 ChaCha20 ops in 3.01s
Doing ChaCha20 ops for 3s on 1024 size blocks: 373266 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 8192 size blocks: 58149 ChaCha20 ops in 2.99s
Doing ChaCha20 ops for 3s on 16384 size blocks: 29032 ChaCha20 ops in 2.97s
version: 3.4.0-dev
built on: Fri Apr 19 06:13:13 2024 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -Wa,--noexecstack -Wall -O3 -march=rv64gcv_zba_zbb -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
CPUINFO: N/A
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
ChaCha20         59807.43k   100758.19k   121359.82k   127408.13k   159316.59k   160154.98k
[cyy@archlinux ssl]$
```

Summary of the result:

Chacha20 Implementation | GCC -march | 16 bytes | 64 bytes | 256 bytes | 1024 bytes | 8192 bytes | 16384 bytes
-- | -- | -- | -- | -- | -- | -- | --
C | 64gc | 50707.96k | 76050.08k | 85786.80k | 88555.45k | 89768.38k | 89509.24k
C | 64gcv_zba_zbb | 57314.60k | 99335.40k | 118663.85k | 124539.49k | 126956.89k | 126828.54k
ASM-RVV | 64gc | 50640.85k | 76239.86k | 121777.83k | 127427.18k | 158146.70k | 158523.45k
ASM-RVV | 64gcv_zba_zbb | 59807.43k | 100758.19k | 121359.82k | 127408.13k | 159316.59k | 160154.98k

According to the result, we get chacha20 speed up to 160154.98k on 16384 bytes blocks, which is 26.3% faster compared to C implementation with rv64gcv_zba_zbb optimized by GCC and 78.9% faster compared to C implementation with rv64gc which is the default -march for riscv in GCC.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [x] tests are added or updated

I also `diff` the generated ASM for Zvkb since I updated the Perl script for ASM generation. There are no changes except for indentation.

t8m commented on crypto/chacha/chacha_riscv.c:
```diff
@@ -44,12 +44,21 @@ void ChaCha20_ctr32_zvkb(unsigned char *out, const unsigned char *inp,
+    if (riscv_vlen() >= 128) {
+        if (RISCV_HAS_ZVKB()) {
+            ChaCha20_ctr32_zvkb(out, inp, len, key, counter);
+        } else {
```
Will this new code work on all RISCV64 CPUs where riscv_vlen() returns >=128? I.e., there is no need to add any further condition?

cyyself replied:
I think Yes.

t8m commented:
Also from your performance measurements I see that the performance on short data is severally degraded and the improvement on long data is only slight. Is it really worth it? Should the existing non-vector implementation be called on shorter inputs?

t8m commented on crypto/chacha/chacha_riscv.c:
```diff
@@ -44,12 +44,21 @@ void ChaCha20_ctr32_zvkb(unsigned char *out, const unsigned char *inp,
+        } else {
+            ChaCha20_ctr32_v(out, inp, len, key, counter);
+        }
+    }
+    else {
```
Nit: these should be on a single line

cyyself commented:
> Also from your performance measurements I see that the performance on short data is severally degraded and the improvement on long data is only slight. Is it really worth it?

I think so. Maybe I should do some profiling and optimize the implementation.

>  Should the existing non-vector implementation be called on shorter inputs?

I will wait until I have further optimization to make this decision.

cyyself commented:
> Also from your performance measurements I see that the performance on short data is severally degraded and the improvement on long data is only slight. Is it really worth it?

I think so. Maybe I should do some profiling and optimize the implementation.

> Should the existing non-vector implementation be called on shorter inputs?

I will wait until I have further optimization to make this decision.

cyyself commented:
It does not pass the `make test` so I convert this PR to a draft. It was modified based on the zvkb implementation, which already in qemu. But it will not pass the test either. I have submitted issue #24070 to address this and will debug it soon.

kroeckx commented:
OPENSSL_riscvcap seems to be undocumented. Can you add an document like OPENSSL_ia32cap and OPENSSL_s390xcap in another PR?

cyyself commented:
> OPENSSL_riscvcap seems to be undocumented. Can you add an document like OPENSSL_ia32cap and OPENSSL_s390xcap in another PR?

I think OPENSSL_riscvcap should be deprecated in the future. For now, we have [hwprobe syscall in the linux kernel since v6.4](https://github.com/torvalds/linux/commit/00e76e2c6a2bd3976d44d4a1fdd0b7a3c2566607) to get what ISA extensions are supported by every CPU.

tom-cosgrove-arm commented:
> I think OPENSSL_riscvcap should be deprecated in the future

Being able to explicitly enable and disable specific optimisations is very useful. On both x86 and Arm we query for capabilities and use that to set the respective xxxcap variables, which can then be given in the environment to override what is detected. I would recommend against deprecating this.

cyyself commented:
> > I think OPENSSL_riscvcap should be deprecated in the future
> 
> Being able to explicitly enable and disable specific optimisations is very useful. On both x86 and Arm we query for capabilities and use that to set the respective xxxcap variables, which can then be given in the environment to override what is detected. I would recommend against deprecating this.

OK. So the things to do might add hwprobe and use it by default, and allow it to be overridden by `OPENSSL_riscvcap`. Missing documentation for `OPENSSL_riscvcap` should be added either.

cyyself commented:
> Also from your performance measurements I see that the performance on short data is severally degraded and the improvement on long data is only slight. Is it really worth it? Should the existing non-vector implementation be called on shorter inputs?

@t8m 

I have refactored the code from #24097. And updated the benchmark on the Canaan Kendryte K230 platform. It shows no performance degradation compared to pure C implementation, which is compiled by GCC 13.2 with default arguments. And we finally got 90379.61k on 16384 bytes block, which is 43.7% faster than implementing in C on rv64gc and 12.6% faster than compiler optimization. However, slightly worse performance can also be seen on small blocks. Since many users will use OpenSSL provided from their distro package with rv64gc baseline without any extension, it is worth using it.

In addition, Canaan Kendryte K230 is an embedded RISC-V SoC with an in-order pipeline with a very short vector length of 128 bits, like Cortex-A53 with NEON in Arm Ecosystems. We may see better performance in out-of-order cores with longer vector lengths on better CPUs.

cyyself commented:
@JerryShih 

I also provided a diff with your implementation using Zvkb. You can see the code below:

```diff
5c5
< # Copyright 2023-2023 The OpenSSL Project Authors. All Rights Reserved.
---
> # Copyright 2023-2024 The OpenSSL Project Authors. All Rights Reserved.
13a14
> # Copyright (c) 2024, Yangyu Chen <cyy@cyyself.name>
40d40
< # - RISC-V Vector Cryptography Bit-manipulation extension ('Zvkb')
63,65c63,65
< # void ChaCha20_ctr32_zvkb(unsigned char *out, const unsigned char *inp,
< #                          size_t len, const unsigned int key[8],
< #                          const unsigned int counter[4]);
---
> # void ChaCha20_ctr32_v(unsigned char *out, const unsigned char *inp,
> #                       size_t len, const unsigned int key[8],
> #                       const unsigned int counter[4]);
124c124,132
<     @{[vror_vi $D0, $D0, 32 - 16]}
---
>     @{[vsll_vi $V24, $D0, 16]}
>     @{[vsll_vi $V25, $D1, 16]}
>     @{[vsll_vi $V26, $D2, 16]}
>     @{[vsll_vi $V27, $D3, 16]}
>     @{[vsrl_vi $D0, $D0, 32 - 16]}
>     @{[vsrl_vi $D1, $D1, 32 - 16]}
>     @{[vsrl_vi $D2, $D2, 32 - 16]}
>     @{[vsrl_vi $D3, $D3, 32 - 16]}
>     @{[vor_vv $D0, $D0, $V24]}
126c134
<     @{[vror_vi $D1, $D1, 32 - 16]}
---
>     @{[vor_vv $D1, $D1, $V25]}
128c136
<     @{[vror_vi $D2, $D2, 32 - 16]}
---
>     @{[vor_vv $D2, $D2, $V26]}
130c138
<     @{[vror_vi $D3, $D3, 32 - 16]}
---
>     @{[vor_vv $D3, $D3, $V27]}
149c157,165
<     @{[vror_vi $B0, $B0, 32 - 12]}
---
>     @{[vsll_vi $V28, $B0, 12]}
>     @{[vsll_vi $V29, $B1, 12]}
>     @{[vsll_vi $V30, $B2, 12]}
>     @{[vsll_vi $V31, $B3, 12]}
>     @{[vsrl_vi $B0, $B0, 32 - 12]}
>     @{[vsrl_vi $B1, $B1, 32 - 12]}
>     @{[vsrl_vi $B2, $B2, 32 - 12]}
>     @{[vsrl_vi $B3, $B3, 32 - 12]}
>     @{[vor_vv $B0, $B0, $V28]}
151c167
<     @{[vror_vi $B1, $B1, 32 - 12]}
---
>     @{[vor_vv $B1, $B1, $V29]}
153c169
<     @{[vror_vi $B2, $B2, 32 - 12]}
---
>     @{[vor_vv $B2, $B2, $V30]}
155c171
<     @{[vror_vi $B3, $B3, 32 - 12]}
---
>     @{[vor_vv $B3, $B3, $V31]}
174c190,198
<     @{[vror_vi $D0, $D0, 32 - 8]}
---
>     @{[vsll_vi $V24, $D0, 8]}
>     @{[vsll_vi $V25, $D1, 8]}
>     @{[vsll_vi $V26, $D2, 8]}
>     @{[vsll_vi $V27, $D3, 8]}
>     @{[vsrl_vi $D0, $D0, 32 - 8]}
>     @{[vsrl_vi $D1, $D1, 32 - 8]}
>     @{[vsrl_vi $D2, $D2, 32 - 8]}
>     @{[vsrl_vi $D3, $D3, 32 - 8]}
>     @{[vor_vv $D0, $D0, $V24]}
176c200
<     @{[vror_vi $D1, $D1, 32 - 8]}
---
>     @{[vor_vv $D1, $D1, $V25]}
178c202
<     @{[vror_vi $D2, $D2, 32 - 8]}
---
>     @{[vor_vv $D2, $D2, $V26]}
180c204
<     @{[vror_vi $D3, $D3, 32 - 8]}
---
>     @{[vor_vv $D3, $D3, $V27]}
199c223,231
<     @{[vror_vi $B0, $B0, 32 - 7]}
---
>     @{[vsll_vi $V28, $B0, 7]}
>     @{[vsll_vi $V29, $B1, 7]}
>     @{[vsll_vi $V30, $B2, 7]}
>     @{[vsll_vi $V31, $B3, 7]}
>     @{[vsrl_vi $B0, $B0, 32 - 7]}
>     @{[vsrl_vi $B1, $B1, 32 - 7]}
>     @{[vsrl_vi $B2, $B2, 32 - 7]}
>     @{[vsrl_vi $B3, $B3, 32 - 7]}
>     @{[vor_vv $B0, $B0, $V28]}
201c233
<     @{[vror_vi $B1, $B1, 32 - 7]}
---
>     @{[vor_vv $B1, $B1, $V29]}
203c235
<     @{[vror_vi $B2, $B2, 32 - 7]}
---
>     @{[vor_vv $B2, $B2, $V30]}
205c237
<     @{[vror_vi $B3, $B3, 32 - 7]}
---
>     @{[vor_vv $B3, $B3, $V31]}
214,216c246,248
< .globl ChaCha20_ctr32_zvkb
< .type ChaCha20_ctr32_zvkb,\@function
< ChaCha20_ctr32_zvkb:
---
> .globl ChaCha20_ctr32_v
> .type ChaCha20_ctr32_v,\@function
> ChaCha20_ctr32_v:
471c503
< .size ChaCha20_ctr32_zvkb,.-ChaCha20_ctr32_zvkb
---
> .size ChaCha20_ctr32_v,.-ChaCha20_ctr32_v
```

You can also take a look at this since this code is changed from your implementation.

ZenithalHourlyRate commented:
> I also provided a diff with your implementation using Zvkb

@cyyself, as the difference is only on the `vror` instruction and all other parts are reused, I think reuse all them in a file named `chacha20-riscv64-v-common.pl` and make extension specific implementions into `chacha20-riscv64-v.pl` and `chacha20-riscv64-zvkb.pl` would largely reduce the maintainance burden.

> Could somebody knowledgeable of RISCV please review this code? I can provide only a formal approval.

I will check the equivalence of these two asm implementation soon. I had some [RISC-V Vector Extension experience inside OpenSSL](https://github.com/openssl/openssl/pull/18479).

> It would be also appreciated if someone could adopt the riscv build targets as community maintainer(s) as they are currently unadopted.
> https://www.openssl.org/policies/general-supplemental/platforms.html

@t8m I am willing to do so, How could I participate in? I think I had related experience, see [here](https://github.com/openssl/openssl/pulls?q=is%3Apr+author%3AZenithalHourlyRate)

cyyself commented:
> > I also provided a diff with your implementation using Zvkb
> 
> @cyyself, as the difference is only on the `vror` instruction and all other parts are reused, I think reusing all of them in a file named `chacha20-riscv64-v-common.pl` and making extension specific implementions into `chacha20-riscv64-v.pl` and `chacha20-riscv64-zvkb.pl` would largely reduce the maintainance burden.
> 

I will do that. Thanks.

t8m commented:
> > It would be also appreciated if someone could adopt the riscv build targets as community maintainer(s) as they are currently unadopted.
> > https://www.openssl.org/policies/general-supplemental/platforms.html
> 
> @t8m I am willing to do so, How could I participate in? I think I had related experience, see [here](https://github.com/openssl/openssl/pulls?q=is%3Apr+author%3AZenithalHourlyRate)

Please open a pull request against https://github.com/openssl/general-policies/ and to move the relevant platforms from unadopted to community. Similar to for example this pull request: https://github.com/openssl/general-policies/pull/54/files

JerryShih commented on crypto/chacha/asm/chacha-riscv64-v.pl:
```diff
@@ -0,0 +1,508 @@
+    @{[vsll_vi $V24, $D0, 16]}
+    @{[vsll_vi $V25, $D1, 16]}
+    @{[vsll_vi $V26, $D2, 16]}
+    @{[vsll_vi $V27, $D3, 16]}
```
I would like to have additional register parameters for `chacha_quad_round_group()` and use them instead of using v-reg id directly.
 
```
    my (
        $A0, $B0, $C0, $D0,
        $A1, $B1, $C1, $D1,
        $A2, $B2, $C2, $D2,
        $A3, $B3, $C3, $D3,

        $T0, $T1, $T2, $T3,   # <=======

        $S_A0, $S_B0, $S_C0, $S_D0,
        $S_A1, $S_B1, $S_C1, $S_D1,
        $S_A2, $S_B2, $S_C2, $S_D2,
        $S_A3, $S_B3, $S_C3, $S_D3,
    ) = @_;

    .....
    @{[vsll_vi $T0, $D0, 16]}
    @{[vsll_vi $T1, $D1, 16]}
    @{[vsll_vi $T2, $D2, 16]}
    @{[vsll_vi $T3, $D3, 16]}
```

cyyself replied:
That is also my concern, I will update them these days.

JerryShih commented on crypto/chacha/build.info:
```diff
@@ -54,3 +54,4 @@ GENERATE[chacha-ia64.S]=asm/chacha-ia64.pl
 GENERATE[chacha-ia64.s]=chacha-ia64.S
 GENERATE[chacha-loongarch64.S]=asm/chacha-loongarch64.pl
 GENERATE[chacha-riscv64-zvkb.s]=asm/chacha-riscv64-zvkb.pl
+GENERATE[chacha-riscv64-v.s]=asm/chacha-riscv64-v.pl
```
I would like to use a single perl generator with the additional argument for `vror` instruction without the new additional perl file.

e.g.
```
GENERATE[chacha-riscv64-zvkb.s]=asm/chacha-riscv64-rvv.pl --enable-zvkb=true
GENERATE[chacha-riscv64-rvv.s]=asm/chacha-riscv64-rvv.pl --enable-zvkb=false
```

cyyself replied:
Thanks. @ZenithalHourlyRate also commented on this.

ZenithalHourlyRate commented:
> I will check the equivalence of these two asm implementation soon.

Reviewed-by: Hongren Zheng \<i@zenithal.me\>

t8m commented:
Please rebase.

cyyself commented:
> Please rebase.

@t8m 

Done. And I also updated the first conversation with updated performance evaluation results.

ZenithalHourlyRate commented on crypto/chacha/asm/chacha-riscv64-zbb-v.pl:
```diff
@@ -305,7 +345,8 @@ sub chacha_quad_round_group {
       $STATE2, $STATE6, $STATE10, $STATE14,
-      $STATE3, $STATE7, $STATE11, $STATE15]}
+      $STATE3, $STATE7, $STATE11, $STATE15,
+      $V24, $V25, $V26, $V27]}
```
just for readability, the comment above `vlsseg_nf_e32_v 8, $V16, $INPUT, $T0` could be `# load the top-half of input data into $V16~$V23` and the comment above `@{[vlsseg_nf_e32_v 8, $V24, $T1, $T0]}` could be `# now load the bottom-half of input data into $V24~$V31` so that readers can more easily infer that `$V24~$V27` at this point are available and can be used as temporary register. A comment above `chacha_quad_round_group` saying `# $V24~$V27 are available now` could be added.

Or more informatively, comments below `map( "v$_", ( 0 .. 31 ) );` could be the usage of each vreg, e.g. `V0~V15 for state, V16~V23 for top-half of input and V24~V31 for bottom-half of input. The vector element of each vreg of the same index, together form the whole state and input (note that vector element width is 32). V24~V27 are available for temporary register at some point. In scalar_data_loop they are used in another way...`

cyyself replied:
Thanks. I added another commit to add these comments, as you suggested.

cyyself commented on crypto/chacha/asm/chacha-riscv64-zbb-v.pl:
```diff
@@ -54,15 +55,18 @@
 my $flavour = $#ARGV >= 0 && $ARGV[0] !~ m|\.|          ? shift : undef;
 
+my $use_zvkb = $flavour && $flavour =~ /zvkb/i ? 1 : 0;
+my $isaext = $use_zvkb ? "_zbb_zvkb" : "_zbb_v";
```
OK. I updated them. Now, I let them both have a `_v_zbb` prefix and when `$use_zvkb` is 1, it will add a `_zvkb` suffix.

JerryShih commented on crypto/chacha/asm/chacha-riscv64-v-zbb.pl:
```diff
@@ -90,6 +94,92 @@
+        @{[vor_vv $C2, $C2, $V_T2]}
+        @{[roriw $S_C2, $S_C2, 32 - $ROL_SHIFT]}
+        @{[vor_vv $C3, $C3, $V_T3]}
+        @{[roriw $S_C3, $S_C3, 32 - $ROL_SHIFT]}
```
We might only have the `$use_zvkb` branch for these code block(the vror code block only).
https://github.com/openssl/openssl/pull/24069/files#diff-10def5210a5379683326679c67ff949d069c2e9eb2a764f03c22f8ef3fa61faaR134-R141

https://github.com/openssl/openssl/pull/24069/files#diff-10def5210a5379683326679c67ff949d069c2e9eb2a764f03c22f8ef3fa61faaR162-R177

cyyself replied:
Thanks for your review. Fixed here: https://github.com/openssl/openssl/pull/24069/commits/f1d7f4daa55bed6e4902ce33a6c5da8858dadea5

JerryShih reviewed (approved):
For me, I would like to have the regular indent in perl generator instead of the final asm file. It's easier for me to maintain the generator.
I'm not sure which one is the good practice in openssl.

cyyself commented:
> For me, I would like to have the regular indent in perl generator instead of the final asm file. It's easier for me to maintain the generator. I'm not sure which one is the good practice in openssl.

I also have this concern. That’s why I separate it to a different commit. I will wait for opinions from OpenSSL maintainers.

JerryShih commented:
> > For me, I would like to have the regular indent in perl generator instead of the final asm file. It's easier for me to maintain the generator. I'm not sure which one is the good practice in openssl.
> 
> I also have this concern. That’s why I separate it to a different commit. I will wait for opinions from OpenSSL maintainers.

The https://github.com/openssl/openssl/commit/f1d7f4daa55bed6e4902ce33a6c5da8858dadea5 has:
1. have branch for vror code block.
2. update indent

It's not just for the indent only.

cyyself commented:
> > For me, I would like to have the regular indent in perl generator instead of the final asm file. It's easier for me to maintain the generator. I'm not sure which one is the good practice in openssl.
> 
> Yes, the indentation should be primarily correct in the .pl file as that is the source file. It is nice if the indentation is correct in the generated file, but it is definitely not a requirement.

Thanks. I edited the last commit on indentation. Now, it only reuses the same part of the code.

t8m commented:
@paulidale please reconfirm

cyyself commented:
Although it has yet to be reviewed for two weeks. There is some good news to share. I got another new RISC-V Board, BananaPi F3, which has Spacemit K1 SoC and 256 bits VLEN, which is 2x wide in Vector Length compared to the K230 I used to benchmark before. The performance is shown below:

```console
[cyy@k1 ssl]$ ./bin/openssl speed -evp chacha20
Doing ChaCha20 ops for 3s on 16 size blocks: 11580221 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 64 size blocks: 4837731 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 256 size blocks: 1433019 ChaCha20 ops in 3.01s
Doing ChaCha20 ops for 3s on 1024 size blocks: 375369 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 8192 size blocks: 47621 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 16384 size blocks: 23843 ChaCha20 ops in 3.01s
version: 3.4.0-dev
built on: Tue Apr 23 06:23:46 2024 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -Wa,--noexecstack -Wall -O3 -march=rv64gc_zba_zbb -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
CPUINFO: N/A
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
ChaCha20         61761.18k   103204.93k   121878.03k   128125.95k   130037.08k   129781.96k
[cyy@k1 ssl]$ export OPENSSL_riscvcap=rv64gc_v_zba_zbb
[cyy@k1 ssl]$ ./bin/openssl speed -evp chacha20
Doing ChaCha20 ops for 3s on 16 size blocks: 11308125 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 64 size blocks: 4804401 ChaCha20 ops in 3.01s
Doing ChaCha20 ops for 3s on 256 size blocks: 1409444 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 1024 size blocks: 725378 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 8192 size blocks: 99129 ChaCha20 ops in 3.00s
Doing ChaCha20 ops for 3s on 16384 size blocks: 51344 ChaCha20 ops in 3.00s
version: 3.4.0-dev
built on: Tue Apr 23 06:23:46 2024 UTC
options: bn(64,64)
compiler: gcc -fPIC -pthread -Wa,--noexecstack -Wall -O3 -march=rv64gc_zba_zbb -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_BUILDING_OPENSSL -DNDEBUG
CPUINFO: N/A
The 'numbers' are in 1000s of bytes per second processed.
type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
ChaCha20         60310.00k   102153.38k   120272.55k   247595.69k   270688.26k   280406.70k
[cyy@k1 ssl]$ neofetch
                   -`                    cyy@k1
                  .o+`                   ------
                 `ooo/                   OS: Arch Linux riscv64
                `+oooo:                  Host: spacemit k1-x deb1 board
               `+oooooo:                 Kernel: 6.1.15-legacy-k1
               -+oooooo+:                Uptime: 6 mins
             `/:-:++oooo+:               Packages: 139 (pacman)
            `/++++/+++++++:              Shell: bash 5.2.26
           `/++++++++++++++:             Terminal: /dev/pts/0
          `/+++ooooooooooooo/`           CPU: Spacemit X60 (8) @ 1.600GHz
         ./ooosssso++osssssso+`          Memory: 111MiB / 3809MiB
        .oossssso-````/ossssss+`
       -osssssso.      :ssssssso.
      :osssssss/        osssso+++.
     /ossssssss/        +ssssooo/-
   `/ossssso+/:-        -:/+osssso+-
  `+sso+:-`                 `.-/+oso:
 `++:.                           `-/+/
 .`                                 `/

[cyy@k1 ssl]$ cat /proc/cpuinfo | head -n 9
processor	: 0
hart		: 0
model name	: Spacemit(R) X60
isa		: rv64imafdcv_sscofpmf_sstc_svpbmt_zicbom_zicboz_zicbop_zihintpause
mmu		: sv39
mvendorid	: 0x710
marchid		: 0x8000000058000001
mimpid		: 0x1000000049772200
```

I changed the baseline to rv64gcv_zba_zbb compiled by GCC. The result shows that it is more than 2x faster than pure C implementation optimized by the compiler. On K230, it only indicates 26.3% faster.

paulidale reviewed (approved):
Apologies, missed this one.

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Merged to the master branch. Thank you for your contribution.
